libeos-updater-util/flatpak.c

```diff
@@ -1294,11 +1301,26 @@ squash_ref_actions_ptr_array (GPtrArray *ref_actions  /* (element-type EuuFlatpa
 
   squashed_ref_actions = g_ptr_array_new_full (g_hash_table_size (hash_table),
                                                (GDestroyNotify) euu_flatpak_remote_ref_action_unref);
-  g_hash_table_iter_init (&hash_iter, hash_table);
 
-  while (g_hash_table_iter_next (&hash_iter, &key, &value))
-    g_ptr_array_add (squashed_ref_actions,
-                     euu_flatpak_remote_ref_action_ref (value));
+  /* Now iterate over the original array so we maintain its order while adding
+   * to the squashed array
+   */
+  for (i = 0; i < ref_actions->len; ++i)
+    {
+      EuuFlatpakRemoteRefAction *action = g_ptr_array_index (ref_actions, i);
+      EuuFlatpakRemoteRefAction *squashed_action_for_ref = NULL;
+
+      if (g_ptr_array_find_with_equal_func (squashed_ref_actions,
+                                            action,
+                                            euu_flatpak_remote_ref_action_equal_ref_only,
+                                            NULL))
+        continue;
+
+      squashed_action_for_ref = g_hash_table_lookup (hash_table, action->ref->ref);
+      g_assert (squashed_action_for_ref != NULL);
+      g_ptr_array_add (squashed_ref_actions,
+                       euu_flatpak_remote_ref_action_ref (squashed_action_for_ref));
```

Comment:
I think this will produce an array which is sometimes not actually the same order as the input array, but I can’t think of a situation where that would cause a problem. It makes me nervous nonetheless.

An example of this would be where the same ref shows up twice in the input list, with two actions, and actions on some other ref in between. For example:
```
ref_actions = [ install A, install B, uninstall A ]
```
will get compressed to the following with this PR:
```
squashed_ref_actions = [ uninstall A, install B ]
```
but I think the original intent of the `squash_ref_actions_ptr_array()` function was to produce:
```
squashed_ref_actions = [ install B, uninstall A ]
```

What do you think?

Reply:
Good point. I doubt that could cause a problem in practice, but it seems worth fixing anyway.

I think there's another potential issue here which is that if in your example A is a dependency of B, the installation of B will fail with either of the two sequences of squashed actions. That seems unlikely to happen since we'd have to add an action which uninstalls a runtime needed by a previous action. I don't think we need to address that here but I added a comment to https://phabricator.endlessm.com/T23150

Pushed fixups to address this and add a unit test. Also spent a while thinking about whether a dependency could be installed too late if it's needed by more than one action, but it looks like the later dependency will be ignored by determine_action_for_related_ref().

Reply:
woohoo, unit tests! Would be good to add a couple more unit tests along the same lines to check the other ways that actions could be squashed and end up out of order. The test you added checks the case where an uninstall action squashes an install action, but there are also the situations:
 * Install squashes uninstall
 * Update squashes update
 * Install squashes update
 * Uninstall squashes updates

Reply:
Honestly I don't think we need a unit test for every one of those. You'd have to make pretty obviously wrong changes in order to break those cases without breaking the test we have now, and each test we add does have some cost (to write and to maintain).

Reply:
> You'd have to make pretty obviously wrong changes in order to break those cases without breaking the test we have now

Well, you made some fairly subtle changes in this PR to go from not working for all the ways actions could be squashed, to working for more of them. So I’d argue that equally subtle changes could break them. The test harness is set up to allow easily adding more actions to a test, as you’ve already done to check the pointer comparison fix.

Reply:
But we are testing for the way in which this was subtly broken (namely, the case described in https://github.com/endlessm/eos-updater/pull/262#discussion_r355992045).

The order preservation code in squash_ref_actions_ptr_array() doesn't do anything that's conditional on action types, so if we added more tests for each type we'd just be re-testing the squashing code that's already covered by other tests.

Anyway, seems like we both agree it's not important either way, so merging this.